task/mgmt-gateway/src/mgs_gimlet.rs

```diff
@@ -543,3 +643,225 @@ fn claim_sp_to_mgs_usart_buf_static(
+            .task
+            .get_mux()
+            .map_err(|err| ResponseError::UpdateFailed(err as u32))?
+        {
+            HfMuxState::SP => (),
+            HfMuxState::HostCPU => return Err(ResponseError::UpdateSlotBusy),
+        }
+
+        // Is an update already in progress?
+        self.buf.ensure_no_update_in_progress()?;
+
+        // Swap to the chosen slot.
+        self.task
+            .set_dev(slot)
+            .map_err(|err| ResponseError::UpdateFailed(err as u32))?;
+
+        // What is the total capacity of the device?
+        let capacity = self
+            .task
+            .capacity()
+            .map_err(|err| ResponseError::UpdateFailed(err as u32))?;
+
+        // How many total sectors do we need to erase? For gimlet, we know that
+        // capacity is an exact multiple of the sector size, which is probably
+        // a safe assumption for future parts as well. We'll assert here in case
+        // that ever becomes untrue, and we can update our math.
+        assert!(capacity % SECTOR_SIZE_BYTES == 0);
+        self.sector_erase.start(capacity / SECTOR_SIZE_BYTES);
+
+        self.buf.start(update.stream_id, update.total_size as usize);
+
+        Ok(())
+    }
+
+    fn prepare_status(
+        &self,
+        request: UpdatePrepareStatusRequest,
+    ) -> Result<UpdatePrepareStatusResponse, ResponseError> {
+        self.buf.ensure_matching_stream_id(request.stream_id)?;
+
+        // Have we failed erasing sectors?
+        if let Some(err) = self.sector_erase.most_recent_error() {
+            return Err(ResponseError::UpdateFailed(err as u32));
+        }
+
+        // We have an update in progress that matches request.stream_id; do we
+        // still have sectors to erase?
+        Ok(UpdatePrepareStatusResponse {
+            done: !self.needs_sectors_erased(),
+        })
+    }
+
+    fn ingest_chunk(
+        &mut self,
+        chunk: UpdateChunk,
+        data: &[u8],
+    ) -> Result<(), ResponseError> {
+        // Have we finished erasing the host flash?
+        if self.needs_sectors_erased() {
+            return Err(ResponseError::UpdateNotPrepared);
+        }
+
+        match self.buf.ingest_chunk(
+            chunk.stream_id,
+            &self.task,
+            chunk.offset,
+            data,
+        )? {
+            UpdateProgress::Complete => {
+                // Update complete; we can now accept a new update.
+                self.buf.reset();
+            }
+            UpdateProgress::Incomplete => (),
+        }
+        Ok(())
+    }
+
+    fn abort(&mut self) -> Result<(), ResponseError> {
+        // TODO should we erase the slot?
+        // TODO should we set_dev() back to what it was (if we changed it)?
+        self.buf.reset();
+        self.sector_erase.abort();
+        Ok(())
+    }
+}
+
+struct HostFlashSectorErase {
+    sectors_to_erase: Range<usize>,
+    most_recent_error: Option<HfError>,
+}
+
+impl Default for HostFlashSectorErase {
```

Comment:
Do we need this `Default` implementation?  It seems a little non-obvious (e.g. I could imagine that the default would be to erase all sectors, instead of none!)

Reply:
Hmm, we need something to the effect of it unless I make `HostFlashUpdate::sector_erase` an `Option`. That might be clearer; I'll give it a shot.